gabxgonz/cat-game-collab
# Install unit tests

# Main changes

This PR adds unit and integration tests for our project. The integration tests are "Play mode" test and the unit tests are "Edit mode" tests.

* We're using the [Unity Test Framework](https://docs.unity3d.com/Packages/com.unity.test-framework@1.1/manual/index.html).
* It uses [NUnit](https://docs.nunit.org/articles/nunit/intro.html) under the hood.
*  I added some Play mode tests asserting the keyboard controls move the player correctly.

## Important Note
`Assets/Scripts` is now an [assembly](https://docs.microsoft.com/en-us/dotnet/standard/assembly/) named "Main". We need this to allow the test assemblies to have access to our game code. 

Our script code previously had global access to all assemblies since it wasn't an assembly itself. Now we need to explicitly give it references to any assemblies it needs.

If you take a look at `Assets/Scripts/Main.asmdef` in the inspector, you'll see references to the InputSystem and TextMeshPro, for instance. Without these references, Unity will throw an error saying it doesn't know what those classes are.

## How to run  tests
* Open the Test Runner frame in Unity by going to `Window > General > Test Runner`
* Select `Play Mode` or `Edit Mode` from the toggle and click the relevant buttons to run the tests!

## How to write tests
* For details on syntaxs, check the docs above.
* Play Mode Tests live in `Assets > Tests > PlayMode`
* Edit Mode Tests live in `Assets > Tests > EditMode`
* Create one test file per source code file per mode
